pandas/core/apply.py

```diff
@@ -299,10 +299,12 @@ def apply_standard(self):
                 result = libreduction.compute_reduction(
                     values, self.f, axis=self.axis, dummy=dummy, labels=labels
                 )
-            except ValueError as err:
-                if "Function does not reduce" not in str(err):
-                    # catch only ValueError raised intentionally in libreduction
-                    raise
+            except ValueError:
+                # TODO there are still cases that give another error (eg GH-31505)
+                # if "Function does not reduce" not in str(err):
+                #     # catch only ValueError raised intentionally in libreduction
+                #     raise
+                pass
```

Comment:
@jbrockmendel I suppose you don't like this "fix", but the reality seems to be that there are other ways that the libreduction can raise an error, that should still be catched. I can try to find a way to avoid calling libreduction (add a case to the `if` check above), but I have the feeling that this would just be a check for the specifc case reported, and there will then just be other cases that raise other errors.

Reply:
So the reason it is now getting here, is because it is object dtype (we already check for extension dtypes and datetimelike dtypes to not get here). Maybe in general object dtype should not go through the cython code?

Reply:
stepping back for a second, it isn't clear to me why it's trying to create a DatetimeTZBlock instead of an ObjectBlock

Reply:
Yep. I tried to step into with the debugger, but didn't directly find the cause. It seems there is some inference at some step, which infers that it is datetimetz dtype, but then the values are not converted and still objects, which then causes an error in creating the block.

Reply:
At some point, it created a DatetimeTZBlock with `array([None], dtype=object)` as its values. It is then copying that block that fails, but the question is how it created such a block in the first place

Reply:
> but the question is how it created such a block in the first place

It looks like in Reducer.get_result self.typ is Series, so we're calling the Series constructor at some point and getting type inference there.  If I add in libreduction L117 `dtype=arr.dtype` tot he `self.typ(...)` call the tests this adds passes (haven't run the rest of the test suite with this edit)

Reply:
Is it safe to assume that reductions are instances of the dtype there, or do we need inference? Naively, it seems like that would fail of `.mean()` on integer dtypes, since that would return a float.

Reply:
> Is it safe to assume that reductions are instances of the dtype there, or do we need inference?

I _think_ this is creating a Series that is effectively a view on some portion of our original Series, so it should be OK.  Definitely merits more eyeballs.

Do we have an idea how big the gains from this are?  this piece of code is disproportionately difficult to maintain.

Reply:
> I think this is creating a Series that is effectively a view on some portion of our original Series, so it should be OK. Definitely merits more eyeballs.

Yep, I think you're correct.

Reply:
@jorisvandenbossche can you try the suggestion in https://github.com/pandas-dev/pandas/pull/31614/files#r374228865 to pass `dtype=self.dtype` to that constructor?

Reply:
It's indeed only creating a subset of the original data, so the dtype should always be correct, and so avoids dtype inference (the calculation of the function only happens afterwards, and thus also a potential dtype change like int->float for mean)